### PR TITLE
Implement SystemChrome.getPreferredOrientations

### PR DESCRIPTION
**Pull request by foreg**
Implement SystemChrome.getPreferredOrientations as requested in https://github.com/flutter/flutter/issues/95719.

getPreferredOrientations returns a list of device orientations as they were requested in setPreferredOrientations. PR in the flutter repo will be created afterwards.

https://github.com/flutter/flutter/issues/95719

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/android/test/io/flutter/plugin/platform/PlatformPluginTest.java
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/plugin/platform/PlatformPluginTest.java:570]: This formatting looks odd can this for loop fit on one line if you make the variable name shorter? 
Can you please add a comment describing where 0 and 15 come from? I think it is 1111 in bits but if you are referring to binary values it might be better to us hex.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java
Comment (foreg) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java:240]: Thank you for the review!

I changed the name from `orientation` to `deviceOrientation`, but I'm not entirely sure that I understood you clearly with this comment. If I did not change the right lines, could you please specify which lines need to be changed?

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java
Comment (foreg) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: Yes, it aligns well with moving integer constants to the DeviceOrientation enum.
In the last commit I use the new encodedBit field, therefore I had to change a `switch` to an `if` statement - is it OK, or do I need to find a better solution?

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java
Comment (foreg) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: Agreed, it helped a lot

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java
Comment (foreg) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java]: Tried to add an explanation why flutter can pass more. Please let me know if it is clear now, or should i rephrase it

Naming is the hardest thing 🥲. `flutterRequestedOrientation` sounds better, thank you!

File: shell/platform/android/test/io/flutter/plugin/platform/PlatformPluginTest.java
Comment (foreg) [reply shell/platform/android/test/io/flutter/plugin/platform/PlatformPluginTest.java]: I introduced the `mostSignificantBit` function which (I hope) helps with the readability here and also removes a "magic" number

**Comment by foreg** [comment]
I accidentally removed [stuartmorgan](https://github.com/stuartmorgan) from the reviewers by clicking the "re-request review" button and have no permissions to bring him back :(

**Comment by reidbaker** [comment]
> I accidentally removed [stuartmorgan](https://github.com/stuartmorgan) from the reviewers by clicking the "re-request review" button and have no permissions to bring him back :(

No worries I added him back :)

**Review by reidbaker — APPROVED** [review]
The andoid parts look good. I still have one open question about if we should give any warning or error if someone passed an incorrect value. 
All this said I am still a new contributor and will defer to @stuartmorgan for their expertise of any of the team conventions I missed during my review.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java
Comment (foreg) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java:475]: It is a good idea, but I'm not sure where is the best place for it:
- in this repo in the `decodeOrientations` method,
- or in the main flutter repo in the `setPreferredOrientations`.

I think the second option is more useful for developers, but it might be an issue to keep that log message updated.

Also, I found the following [issue](https://github.com/flutter/flutter/issues/99149) which tracks the problem.

Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/PlatformChannel.java:475]: I can see updating setPreferredOrientations also being valuable but in this PR I think warning that the decode found something it didnt expect even if the passed value has an issue someplace else. 

Since you agree lets add a warning in `decodeOrientations`.

**Comment by camsim99** [comment]
Note from triage: waiting on review @stuartmorgan

**Comment by stuartmorgan** [comment]
Sorry, I should have redirected this to other reviewers earlier; I'm not familiar with this code, and it should have a web and iOS reviewer for those parts of the code. @reidbaker Did you want me to review the Android part?

**Comment by stuartmorgan** [comment]
(@yjbanov, @jmagman, feel free to delegate to others, I just wanted to make sure we get someone from each affected platform to review the relevant code.)

**Comment by reidbaker** [comment]
> Sorry, I should have redirected this to other reviewers earlier; I'm not familiar with this code, and it should have a web and iOS reviewer for those parts of the code. @reidbaker Did you want me to review the Android part?

No I wanted to make sure all the flutterisms were enforced because I am still not confident in my ability to follow all of them and am not comfortable signing off on the non android parts of this code. 
Review from the others you tagged can fill that role.

File: lib/web_ui/lib/src/engine/embedder.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/embedder.dart:309]: nit: s/orientation/orientations/

File: lib/web_ui/test/engine/window_test.dart
Comment (yjbanov) [inline lib/web_ui/test/engine/window_test.dart:431]: This test doesn't seem to use this variable for anything meaningful. It just puts stuff in it then clears it.

File: lib/web_ui/test/engine/window_test.dart
Comment (yjbanov) [inline lib/web_ui/test/engine/window_test.dart:432]: Same for this variable.

File: lib/web_ui/test/engine/window_test.dart
Comment (yjbanov) [inline lib/web_ui/test/engine/window_test.dart:529]: Is this the desired outcome if we failed to lock onto `portraitDown`?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:184]: Since you already need to handle each case in the switch anyway, how about avoiding the loop and c-style array and length:
```suggestion
  if (supportedOrientations & UIInterfaceOrientationMaskPortrait) {
    [orientations addObject:@"DeviceOrientation.portraitUp"];
  }
  if (supportedOrientations & UIInterfaceOrientationMaskPortraitUpsideDown) {
    [orientations addObject:@"DeviceOrientation.portraitDown"];
  }
  if (supportedOrientations & UIInterfaceOrientationMaskLandscapeLeft) {
    [orientations addObject:@"DeviceOrientation.landscapeLeft"];
  }
  if (supportedOrientations & UIInterfaceOrientationMaskLandscapeRight) {
    [orientations addObject:@"DeviceOrientation.landscapeRight"];
  }
```

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:158]: > getPreferredOrientations returns a list of device orientations as they were requested in setPreferredOrientations

Does it actually do this?  It seems like it's ignoring whatever was previously set and is instead always returning the UIViewController supported orientations.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:68]: This needs a test in https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterPlatformPluginTest.mm.  You should be able to do something like:
```
  FlutterEngine* engine = [[[FlutterEngine alloc] initWithName:@"test" project:nil] autorelease];
  [engine runWithEntrypoint:nil];
  FlutterViewController* flutterViewController =
      [[FlutterViewController alloc] initWithEngine:engine nibName:nil bundle:nil];

  id flutterViewControllerMock = OCMPartialMock(flutterViewController);
  OCMStub([flutterViewControllerMock supportedInterfaceOrientations]).andReturn(UIInterfaceOrientationMaskPortraitUpsideDown & UIInterfaceOrientationMaskLandscapeRight);
  engine.viewController = flutterViewControllerMock;

  std::unique_ptr<fml::WeakPtrFactory<FlutterEngine>> _weakFactory =
      std::make_unique<fml::WeakPtrFactory<FlutterEngine>>(engine);
  FlutterPlatformPlugin* plugin =
      [[[FlutterPlatformPlugin alloc] initWithEngine:_weakFactory->GetWeakPtr()] autorelease];
```
then call the `handleMethodCall` and check the result (there are some examples in that test file I think).

**Review by jmagman — CHANGES_REQUESTED** [review]
Meant to request changes

File: lib/web_ui/test/engine/window_test.dart
Comment (foreg) [reply lib/web_ui/test/engine/window_test.dart:529]: Thank you for the comments! I will fix those three above in the following commit.

Yes, I believe `GetPreferredOrientations` should return requested orientations without any changes. Please, consider the following case:
```dart
// somewhere in my app
final requested = [DeviceOrientation.portraitDown];
await SystemChrome.setPreferredOrientations(requested);

// later in a third party package
final preferredByMainApp = await SystemChrome.getPreferredOrientations();
final requestedByPackage = [DeviceOrientation.portraitUp];
await SystemChrome.setPreferredOrientations(requestedByPackage);

// the package can restore preferred orientations
await SystemChrome.setPreferredOrientations(preferredByMainApp);
```
I thought, if `getPreferredOrientations` will return an empty list, than the last call to `setPreferredOrientations` wouldn't be the same as the first one - it seems wrong to me.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm
Comment (foreg) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:158]: Thanks for the review! I really appreciated code snippets, because i'm not confident in my obj-c skills 😅

Hmm, I'm making the assumption that requested orientations are stored in `_orientationPreferences` (inside `FlutterViewController`) and returned in `supportedInterfaceOrientations`. 

Now I'm thinking about adding a public field `flutterRequestedOrientation` in the `FlutterViewController` and using it in the`getSystemChromePreferredOrientations`

File: lib/web_ui/lib/src/engine/embedder.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/embedder.dart:312]: Looks like this should be in plural form?
```suggestion
  Future<List<dynamic>> getPreferredOrientations() async {
```

**Review by mdebbar — COMMENTED** [review]
Changes in `lib/web_ui` look good to me once @yjbanov's comments are addressed.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:158]: Is there any point to putting this in the embedder then?  Can't that just be tracked in the dart code?

Comment (foreg) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:158]: Yes, now it looks like an over-complicated logic, compared to a simple solution that I filed to the framework repo - https://github.com/flutter/flutter/pull/123474. I feel awkward for wasting your time, although I really appreciate all of your comments

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:158]: No problem at all, thank you for thinking it all through!

Does this mean this PR should be closed in favor of https://github.com/flutter/flutter/pull/123474?

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:158]: Looks like https://github.com/flutter/flutter/pull/123474 was closed, so it sounds like this should be too for the same reasons?  You may want to start with a concrete design proposal that can be iterated on.

**Comment by jmagman** [comment]
> Yes, now it looks like an over-complicated logic, compared to a simple solution that I filed to the framework repo - https://github.com/flutter/flutter/pull/123474.

> Looks like [flutter/flutter#123474](https://github.com/flutter/flutter/pull/123474) was closed, so it sounds like this should be too for the same reasons? You may want to start with a concrete design proposal that can be iterated on.

I'm going to close this for now based on our above conversation https://github.com/flutter/engine/pull/39763#discussion_r1170475644.  You may want to add more details to https://github.com/flutter/flutter/issues/95719 based on what you've discovered.  Please feel free to open a new PR or reopen this one based on feedback there.

**Pull request closed without merging**
